drivers/rf-transceiver/hmc6300/hmc6300.c

```diff
@@ -0,0 +1,469 @@
+{
+	int ret, e;
+
+	d->vco.freqs = (uint64_t *) calloc(entries, sizeof(uint64_t));
+	if (!d->vco.freqs)
+		return -ENOMEM;
+
+	d->vco.fbdiv = (uint8_t *) calloc(entries, sizeof(uint8_t));
+	if (!d->vco.fbdiv) {
+		ret = -ENOMEM;
+		goto error;
+	}
+
+	for(e = 0; e < entries; e++) {
+		d->vco.freqs[e] = f_start + e * f_step;
+		d->vco.fbdiv[e] = fbdiv_start + e;
+	}
+
+	d->vco.entries = entries;
+
+	return 0;
+error:
+	free(d->vco.freqs);
+	return ret;
+}
+
+/* Create hmc6300 device and initialize it from init params. */
+int hmc6300_init(struct hmc6300_dev **dev, struct hmc6300_init_param *init)
+{
+	struct hmc6300_dev *d;
+	int ret;
+
+	if (!dev || !init)
+		return -EINVAL;
+
+	d = (struct hmc6300_dev *) calloc(1, sizeof(*d));
+	if (!d)
+		return -ENOMEM;
+
+	switch(init->ref_clk) {
+	case HMC6300_REF_CLK_71p42857MHz:
+		ret = _hmc6300_generate_vco(d, 56500000000, 250000000, 0x22, 31);
+		break;
+	case HMC6300_REF_CLK_75MHz:
+		ret = _hmc6300_generate_vco(d, 55125000000, 262500000, 0x12, 43);
+		break;
+	case HMC6300_REF_CLK_142p8571MHz:
+		ret = _hmc6300_generate_vco(d, 56500000000, 500000000, 0x11, 16);
+		break;
+	case HMC6300_REF_CLK_154p2857MHz:
+		ret = _hmc6300_generate_vco(d, 57240000000, 540000000, 0x0a, 17);
+		break;
+	default:
+		ret = -EINVAL;
+	}
+
+	if (ret)
+		goto error_0;
+
+	ret = no_os_gpio_get(&d->en, &init->en);
+	if (ret)
+		goto error_1;
+
+	ret = no_os_gpio_get(&d->clk, &init->clk);
+	if (ret)
+		goto error_2;
+
+	ret = no_os_gpio_get(&d->data, &init->data);
+	if (ret)
+		goto error_3;
+
+	ret = no_os_gpio_get(&d->scanout, &init->scanout);
+	if (ret)
+		goto error_4;
+
+	ret = no_os_gpio_direction_output(d->en, NO_OS_GPIO_HIGH);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_direction_output(d->clk, NO_OS_GPIO_LOW);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_direction_output(d->data, NO_OS_GPIO_LOW);
+	if (ret)
+		goto error;
+
+	ret = no_os_gpio_direction_input(d->scanout);
+	if (ret)
+		goto error;
+
+	ret = hmc6300_write_regmap(d, hmc6300_default_regmap);
+	if (ret)
+		goto error;
+
+	*dev = d;
+	return 0;
+error:
+	no_os_gpio_remove(d->scanout);
+error_4:
+	no_os_gpio_remove(d->data);
+error_3:
+	no_os_gpio_remove(d->clk);
+error_2:
+	no_os_gpio_remove(d->en);
+error_1:
+	free(d->vco.freqs);
+	free(d->vco.fbdiv);
+error_0:
+	free(d);
+	return ret;
+}
+
+/* Remove the hmc6300 device by deallocating resources. */
+int hmc6300_remove(struct hmc6300_dev *dev)
+{
+	free(dev->vco.freqs);
+	free(dev->vco.fbdiv);
+	no_os_gpio_remove(dev->en);
+	no_os_gpio_remove(dev->clk);
+	no_os_gpio_remove(dev->data);
+	no_os_gpio_remove(dev->scanout);
+	free(dev);
+	return 0;
+}
+
+/* Write a device row using GPIO bit-banging. */
+int hmc6300_write_row(struct hmc6300_dev *dev, uint8_t row, uint8_t val)
+{
+	int b;
+	uint32_t send;
+
+	if (!dev)
+		return -EINVAL;
+
+	send = val; // data
+	send |= ((uint32_t)row & 0x3f) << 8; // array address
+	send |= 1 << 14; // write
+	send |= ((uint32_t)0x6) << 15;  // chip address
```

Comment:
The readability may be improved by using `no_os_field_prep()` and macros for the magic values. Also, the casts are not needed.

Reply:
will do